db/hints/manager.cc

```diff
@@ -417,6 +417,27 @@ future<db::commitlog> manager::end_point_hints_manager::add_store() noexcept {
             // in resource_manager, so its redundant for the commitlog to apply
             // a hard limit.
             cfg.allow_going_over_size_limit = true;
+            // The API for waiting for hint replay relies on replay positions
+            // monotonically increasing. When there are no segments on disk,
+            // by default the commitlog will calculate the first segment ID
+            // based on the boot time. This may cause the following sequence
+            // of events to occur:
+            //
+            // 1. Node starts with empty hints queue
+            // 2. Some hints are written and some segments are created
+            // 3. All hints are replayed
+            // 4. Hint sync point is created
+            // 5. Commitlog instance gets re-created and resets it segment ID counter
+            // 6. New hint segment has the first ID as the first (deleted by now) segment
+            // 7. Waiting for the sync point commences but resolves immediately
+            //    before new hints are replayed - since point 5., `_last_written_rp`
+            //    and `_sent_upper_bound_rp` are not updated because RPs of new
+            //    hints are much lower than both of those marks.
+            //
+            // In order to prevent this situation, we override the base segment ID
+            // of the newly created commitlog instance - it should start with an ID
+            // which is larger than the segment ID of the RP of the last written hint.
+            cfg.base_segment_id = _last_written_rp.base_id();
```

Comment:
In the comment you say "larger" - shouldn't it be `_last_written_rp.base_id() + 1` then? Or is it incremented in some other place?

Reply:
The `_ids` field is incremented before using its value is used for the first time:

https://github.com/scylladb/scylla/blob/3f2c680b70d213b8eb76b63f4e54e46225ed9923/db/commitlog/commitlog.cc#L351-L353

By the way, here is the whole logic which initializes `_ids` - so it's actually `max(base_segment_id, max(seg.base_id() for seg : segments_existing_on_disk)`, however in case of hints the custom `base_segment_id` should always be a (possibly non-exclusive) winner in this maximum contest.

https://github.com/scylladb/scylla/blob/3f2c680b70d213b8eb76b63f4e54e46225ed9923/db/commitlog/commitlog.cc#L1299-L1309

Perhaps the parameter should have a better name (I'm open for suggestions), and/or it should be more thoroughly described by a comment (I'll try to do it tomorrow).

Reply:
Added a more detailed description about `base_segment_id`.